File: 2018/day01/src/main.jl
function first_part(freqs)
    total = 0

    for f in split(freqs)
        num = parse.(Int, f)
        total += num
    end

    println(total)
end

function second_part(freqs)
    total = 0
    seen = Set([0])
    twice = -1

    while twice <= -1
        for f in split(freqs)
            num = parse.(Int, f)
            total += num

            if total in seen
                twice = total
                break
            end

            push!(seen, total)
        end
    end

    println(twice)
end

file = open("../input/input.txt")
freqs = read(file, String)

first_part(freqs)
second_part(freqs)

close(file)

File: 2018/day01/src/main.rs
use std::{
    collections::HashSet,
    error::Error,
    io::{ Read, stdin }
};

fn first_part(input: &str) -> Result<(), Box<Error>> {
    let mut total = 0;

    for i in input.lines() {
        let num: i32 = i.parse()?;

        total += num;
    }

    println!("{}", total);

    Ok(())
}

fn second_part(input: &str) -> Result<(), Box<Error>> {
    let mut total = 0;

    let mut seen = HashSet::new();
    seen.insert(0);

    loop {
        for i in input.lines() {
            let num: i32 = i.parse()?;

            total += num;

            if seen.contains(&total) {
                println!("{}", total);
                return Ok(());
            }

            seen.insert(total);
        }
    }
}

fn main() -> Result<(), Box<Error>> {
    let mut input = String::new();
    stdin().read_to_string(&mut input)?;

    first_part(&input)?;
    second_part(&input)?;

    Ok(())
}
